### PR TITLE
Begin implementing the new Gandi APIs

### PR DESCRIPTION
**Pull request by thommay**
Requires tiramiseb/go-gandi#2 ; I've been testing with the
addition of

```
replace github.com/tiramiseb/go-gandi => ../go-gandi
```

in my `go.mod`.

### CODE REVIEW

**Comment by thommay** [comment]
also this is super rough right now, definitely WIP.

**Comment by thommay** [comment]
So this is exciting:
```
Terraform will perform the following actions:

  ~ gandi_domain.xxxx_dev
      nameservers.#: "3" => "4"
      nameservers.0: "ns-38-a.gandi.net" => "ns-70.awsdns-08.com"
      nameservers.1: "ns-203-b.gandi.net" => "ns-1153.awsdns-16.org"
      nameservers.2: "ns-230-c.gandi.net" => "ns-936.awsdns-53.net"
      nameservers.3: "" => "ns-1604.awsdns-08.co.uk"


Plan: 0 to add, 1 to change, 0 to destroy.
```

**Comment by tiramiseb** [comment]
Do we really need IDE-specific files? (I'm thinking about the `.idea` folder). Ignoring `/.idea` would be okay, but having files in it...

**Comment by tiramiseb** [comment]
I have not checked it works correctly, I have currently stopped using Terraform. I trust you with that.

**Comment by thommay** [comment]
yeah, i can ignore the `.idea` directory

**Comment by thommay** [comment]
@baloo Does Gandi support/could Gandi provide a test org that we could  test this plugin against without having to actually buy any number of domains? The Dry Run mode helps a bit but it's hard to validate the whole flow with.

**Comment by baloo** [comment]
I do not believe we have that publicly available, I'll see what I can do.

**Comment by thommay** [comment]
> I do not believe we have that publicly available, I'll see what I can do.

@baloo any news or ideas for this?

**Comment by baloo** [comment]
Not much. I thought we could use a demo account but it needs some work. Nothing available in the short term :(

**Comment by thommay** [comment]
@tiramiseb this is ready to merge.

**Comment by tiramiseb** [comment]
Sorry, I prepared a review days ago, but I forgot to submit it... :sweat_smile:

File: gandi/resource_domain.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		}
+		d.SetPartial("nameservers")
+	}
+	d.Partial(false)
+	return resourceDomainRead(d, m)
+}
+
+// The Gandi API doesn't presently support deleting domains
+func resourceDomainDelete(d *schema.ResourceData, m interface{}) error {
+	d.SetId("")
```
[the remark this answers is not in the document]
Comment (thommay) [reply gandi/resource_domain.go]: Yeah, this is the right way of doing it - terraform requires the delete operation to be implemented, and all that we can do is remove the object from the state.

File: gandi/structures_domain.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	m["email"] = in.Email
+	m["family_name"] = in.FamilyName
+	m["given_name"] = in.GivenName
+	m["street_addr"] = in.StreetAddr
+	m["type"] = flattenContactType(in.ContactType)
+
+	return []interface{}{m}
+}
+
+func flattenContactType(cnt int) (ret string) {
```
[the remark this answers is not in the document]
Comment (thommay) [reply gandi/structures_domain.go]: i think i prefer the explicit mapping in the function, but yeah, not _that_ bothered either way.

[the remark this answers is not in the document]
Comment (tiramiseb) [reply gandi/structures_domain.go]: I have the feeling it is better, performance-wise, to have a "global" map (or even a list in that case), instantiated only once, than a function containing a switch/case structure...

Maybe I'm wrong, dunno...

**Comment by thommay** [comment]
> Sorry, I prepared a review days ago, but I forgot to submit it...

No worries, happens to all of us. @tiramiseb I've fixed your comments.

**Pull request merged**